### PR TITLE
API for Email Campaign feature implementation

### PR DESCRIPTION
**Pull request by ippezshelby**
@satish-conflux could you please find time and review this so that its can be a complement for sms campaign. And also have the UI implementation for this.

### CODE REVIEW

File: fineract-provider/dependencies.gradle [hunk 1]
```diff
@@ -90,7 +90,7 @@ dependencies {
                 // it's useful to have this for the Spring Boot TestRestTemplate http://docs.spring.io/spring-boot/docs/current-SNAPSHOT/reference/htmlsingle/#boot-features-rest-templates-test-utility
                 [group: 'org.apache.httpcomponents', name: 'httpclient', version: '4.3.5'],
                 // Once we've switched to Java 8 this dep can be removed.
-                //[group: 'com.google.code.findbugs', name: 'jsr305', version: '3.0.0']
+                [group: 'com.google.code.findbugs', name: 'jsr305', version: '3.0.0']
```
Comment (nazeer1100126) [inline fineract-provider/dependencies.gradle:93]: findbugs doesn't fit with Apache License.

File: fineract-provider/src/main/java/org/apache/fineract/infrastructure/campaigns/email/api/EmailApiResource.java [hunk 2]
```diff
@@ -0,0 +1,188 @@
...
+    private final String resourceNameForPermissions = "Email";
+    private final PlatformSecurityContext context;
+    private final EmailReadPlatformService readPlatformService;
+    private final DefaultToApiJsonSerializer<EmailData> toApiJsonSerializer;
+    private final ApiRequestParameterHelper apiRequestParameterHelper;
+    private final PortfolioCommandSourceWritePlatformService commandsSourceWritePlatformService;
+    private final EmailConfigurationReadPlatformService emailConfigurationReadPlatformService;
+
+    @Autowired
+    public EmailApiResource(final PlatformSecurityContext context, final EmailReadPlatformService readPlatformService,
```
Comment (nazeer1100126) [inline fineract-provider/src/main/java/org/apache/fineract/infrastructure/campaigns/email/api/EmailApiResource.java:64]: Why EmailApiResource is having emailcampaign as end point? It should emails right?

File: fineract-provider/src/main/java/org/apache/fineract/infrastructure/campaigns/email/api/EmailCampaignApiResource.java [hunk 3]
```diff
@@ -0,0 +1,229 @@
...
+import java.util.HashSet;
+
+/**
+ * Created with IntelliJ IDEA.
+ * User: andrew
+ * Date: 19-5-14
+ * Time: 15:17
+ * To change this template use File | Settings | File Templates.
+ */
+@Path("/emailcampaign/campaign")
```
Comment (nazeer1100126) [inline fineract-provider/src/main/java/org/apache/fineract/infrastructure/campaigns/email/api/EmailCampaignApiResource.java:56]: Path should be emailcampaigns

File: fineract-provider/src/main/java/org/apache/fineract/infrastructure/campaigns/email/api/EmailApiResource.java [hunk 4]
```diff
@@ -0,0 +1,188 @@
...
+        this.context = context;
+        this.readPlatformService = readPlatformService;
+        this.toApiJsonSerializer = toApiJsonSerializer;
+        this.apiRequestParameterHelper = apiRequestParameterHelper;
+        this.commandsSourceWritePlatformService = commandsSourceWritePlatformService;
+        this.emailConfigurationReadPlatformService = emailConfigurationReadPlatformService;
+    }
+
+    @GET
+    public String retrieveAll(@Context final UriInfo uriInfo) {
```
Comment (nazeer1100126) [inline fineract-provider/src/main/java/org/apache/fineract/infrastructure/campaigns/email/api/EmailApiResource.java:77]: Implement pagination for better memory and process utilization

File: fineract-provider/src/main/java/org/apache/fineract/infrastructure/campaigns/email/api/EmailApiResource.java [hunk 5]
```diff
@@ -0,0 +1,188 @@
...
+
+        final Collection<EmailData> emailMessages = this.readPlatformService.retrieveAll();
+
+        final ApiRequestJsonSerializationSettings settings = this.apiRequestParameterHelper.process(uriInfo.getQueryParameters());
+        return this.toApiJsonSerializer.serialize(settings, emailMessages);
+    }
+
+    @GET
+    @Path("pendingEmail")
+    public String retrievePendingEmail(@Context final UriInfo uriInfo,@QueryParam("limit") final Long limit) {
```
Comment (nazeer1100126) [inline fineract-provider/src/main/java/org/apache/fineract/infrastructure/campaigns/email/api/EmailApiResource.java:89]: emails by status API is sufficient. Is this API required? if yes please provide offset and limit params for pagination

File: fineract-provider/src/main/java/org/apache/fineract/infrastructure/campaigns/email/api/EmailApiResource.java [hunk 6]
```diff
@@ -0,0 +1,188 @@
...
+
+        final Collection<EmailData> emailMessages = this.readPlatformService.retrieveAllPending(limit.intValue());
+
+        final ApiRequestJsonSerializationSettings settings = this.apiRequestParameterHelper.process(uriInfo.getQueryParameters());
+        return this.toApiJsonSerializer.serialize(settings, emailMessages);
+    }
+    
+    @GET
+    @Path("sentEmail")
+    public String retrieveSentEmail(@Context final UriInfo uriInfo, @QueryParam("limit") final Long limit) {
```
Comment (nazeer1100126) [inline fineract-provider/src/main/java/org/apache/fineract/infrastructure/campaigns/email/api/EmailApiResource.java:101]: emails by status API is sufficient. Is this API required? if yes please provide offset and limit params for pagination

File: fineract-provider/src/main/java/org/apache/fineract/infrastructure/campaigns/email/api/EmailApiResource.java [hunk 7]
```diff
@@ -0,0 +1,188 @@
...
+
+        final Page<EmailData> emailMessages = this.readPlatformService.retrieveEmailByStatus(limit.intValue(),status.intValue(),fromDate,toDate);
+
+        final ApiRequestJsonSerializationSettings settings = this.apiRequestParameterHelper.process(uriInfo.getQueryParameters());
+        return this.toApiJsonSerializer.serialize(settings, emailMessages);
+    }
+    
+    @GET
+    @Path("failedEmail")
+    public String retrieveFailedEmail(@Context final UriInfo uriInfo, @QueryParam("limit") final Long limit) {
```
Comment (nazeer1100126) [inline fineract-provider/src/main/java/org/apache/fineract/infrastructure/campaigns/email/api/EmailApiResource.java:136]: emails by status API is sufficient. Is this API required? if yes please provide offset and limit params for pagination

File: fineract-provider/src/main/java/org/apache/fineract/infrastructure/campaigns/email/api/EmailCampaignApiResource.java [hunk 8]
```diff
@@ -0,0 +1,229 @@
...
+
+        EmailCampaignData emailCampaignData = this.emailCampaignReadPlatformService.retrieveOne(resourceId);
+        final ApiRequestJsonSerializationSettings settings = this.apiRequestParameterHelper.process(uriInfo.getQueryParameters());
+        return this.emailCampaignDataDefaultToApiJsonSerializer.serialize(settings,emailCampaignData);
+
+    }
+
+    @GET
+    @Produces({ MediaType.APPLICATION_JSON })
+    public String retrieveAllCampaign(@Context final UriInfo uriInfo) {
```
Comment (nazeer1100126) [inline fineract-provider/src/main/java/org/apache/fineract/infrastructure/campaigns/email/api/EmailCampaignApiResource.java:116]: Better to have pagination

File: fineract-provider/src/main/java/org/apache/fineract/infrastructure/campaigns/email/api/EmailCampaignApiResource.java [hunk 9]
```diff
@@ -0,0 +1,229 @@
...
+        campaignMessage = this.emailCampaignWritePlatformService.previewMessage(query);
+        final ApiRequestJsonSerializationSettings settings = this.apiRequestParameterHelper.process(uriInfo.getQueryParameters());
+        return this.previewCampaignMessageDefaultToApiJsonSerializer.serialize(settings,campaignMessage, new HashSet<String>());
+
+    }
+
+
+    @GET()
+    @Path("template")
+    public String retrieveAll(@Context final UriInfo uriInfo){
```
Comment (nazeer1100126) [inline fineract-provider/src/main/java/org/apache/fineract/infrastructure/campaigns/email/api/EmailCampaignApiResource.java:194]: the method name should reflect the purpose. In this case it is not retrieving all campaigns.

File: fineract-provider/src/main/java/org/apache/fineract/commands/service/CommandWrapperBuilder.java [hunk 10]
```diff
@@ -2803,4 +2803,82 @@ public CommandWrapperBuilder deleteReportMailingJob(final String entityName, fin
         this.href = "/reportmailingjobs/" + entityId;
         return this;
     }
+	
+	    public CommandWrapperBuilder createEmail() {
+        this.actionName = "CREATE";
+        this.entityName = "EMAIL";
+        this.entityId = null;
+        this.href = "/emailcampaign/template";
```
Comment (nazeer1100126) [inline fineract-provider/src/main/java/org/apache/fineract/commands/service/CommandWrapperBuilder.java:2811]: change all href to actual end point name.

File: fineract-provider/src/main/java/org/apache/fineract/infrastructure/campaigns/email/domain/EmailCampaign.java [hunk 11]
```diff
@@ -0,0 +1,538 @@
...
+import org.apache.fineract.useradministration.domain.AppUser;
+import org.springframework.data.jpa.domain.AbstractPersistable;
+
+import javax.persistence.*;
+import java.security.InvalidParameterException;
+import java.util.*;
+
+@Entity
+@Table(name = "scheduled_email_campaign")
+public class EmailCampaign extends AbstractPersistable<Long> {
```
Comment (nazeer1100126) [inline fineract-provider/src/main/java/org/apache/fineract/infrastructure/campaigns/email/domain/EmailCampaign.java:44]: SMSCampaign and EmailCampaign having common instance variables. Can we move them to mapped super class and extend sms and email campaign from that?

File: fineract-provider/src/main/java/org/apache/fineract/infrastructure/campaigns/email/domain/ScheduledEmailStretchyReportParamDateOption.java [hunk 12]
```diff
@@ -0,0 +1,117 @@
...
+ * KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.fineract.infrastructure.campaigns.email.domain;
+
+public enum ScheduledEmailStretchyReportParamDateOption {
+    INVALID(0, "scheduledEmailStretchyReportParamDateOption.invalid", "invalid"),
+    TODAY(1, "scheduledEmailStretchyReportParamDateOption.today", "today"),
+    YESTERDAY(2, "scheduledEmailStretchyReportParamDateOption.yesterday", "yesterday"),
```
Comment (nazeer1100126) [inline fineract-provider/src/main/java/org/apache/fineract/infrastructure/campaigns/email/domain/ScheduledEmailStretchyReportParamDateOption.java:24]: In which case YESTERDAY will be used?

File: fineract-provider/src/main/java/org/apache/fineract/infrastructure/campaigns/email/handler/UpdateEmailCommandHandler.java [hunk 13]
```diff
@@ -0,0 +1,47 @@
...
+import org.apache.fineract.infrastructure.core.api.JsonCommand;
+import org.apache.fineract.infrastructure.core.data.CommandProcessingResult;
+import org.apache.fineract.infrastructure.campaigns.email.service.EmailWritePlatformService;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.stereotype.Service;
+import org.springframework.transaction.annotation.Transactional;
+
+@Service
+@CommandType(entity = "EMAIL", action = "UPDATE")
+public class UpdateEmailCommandHandler implements NewCommandSourceHandler {
```
Comment (nazeer1100126) [inline fineract-provider/src/main/java/org/apache/fineract/infrastructure/campaigns/email/handler/UpdateEmailCommandHandler.java:32]: What is the use case to update an email?

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/client/domain/ClientRepository.java [hunk 14]
```diff
@@ -21,6 +21,6 @@
 import org.springframework.data.jpa.repository.JpaRepository;
 import org.springframework.data.jpa.repository.JpaSpecificationExecutor;
 
-abstract interface ClientRepository extends JpaRepository<Client, Long>, JpaSpecificationExecutor<Client> {
+public abstract interface ClientRepository extends JpaRepository<Client, Long>, JpaSpecificationExecutor<Client> {
```
Comment (nazeer1100126) [inline fineract-provider/src/main/java/org/apache/fineract/portfolio/client/domain/ClientRepository.java:24]: we are not exposing ClientRepository. We need to make use of ClientRepositoryWrapper

File: fineract-provider/src/main/resources/sql/migrations/core_db/V322_2__email_business_rules.sql [hunk 15]
```diff
@@ -0,0 +1,127 @@
+--
```
Comment (nazeer1100126) [inline fineract-provider/src/main/resources/sql/migrations/core_db/V322_2__email_business_rules.sql:1]: Most of the stretchy reports are common with SMS Campaigns

**Comment by nazeer1100126** [comment]
I see we can make use of existing SMS Campaign classes with little tweaks.

**Comment by ippezshelby** [comment]
Hi @nazeer1100126 , Could you please make the little tweaks with the existing SMS Campaign if you are not so busy or you can delegate @satish-conflux  to work on these changes as well please.. Thanks

**Comment by nazeer1100126** [comment]
@Ippezrobert Myself and Satish are working on other tasks. So we can't work on this.

**Comment by ippezshelby** [comment]
Closing and Sending New PR

**Pull request closed without merging**